File: backend/src/services/AppSettingService.ts
import fs from 'node:fs';
import path from 'node:path';
import cron from 'node-cron';
import { prisma } from '../db/client.js';
import { logger } from '../config/logger.js';

const DEFAULT_BACKUP_DIR = process.env.DB_BACKUP_DIR || path.join(process.cwd(), 'db-backup');
const DEFAULT_BACKUP_CRON = '0 19 * * 5';

export class AppSettingService {
  async getSetting(key: string): Promise<string | null> {
    const row = await prisma.appSetting.findUnique({ where: { key } });
    return row?.value ?? null;
  }

  async setSetting(key: string, value: string): Promise<void> {
    await prisma.appSetting.upsert({
      where: { key },
      update: { value, updatedAt: BigInt(Date.now()) },
      create: { key, value, updatedAt: BigInt(Date.now()) },
    });
  }

  /**
   * 获取备份目录：优先 DB 配置，fallback env var，再 fallback cwd/db-backup
   */
  async getBackupDir(): Promise<string> {
    const dbValue = await this.getSetting('backup_dir');
    return dbValue?.trim() || DEFAULT_BACKUP_DIR;
  }

  /**
   * 获取备份 cron 表达式：优先 DB 配置，fallback env var，再 fallback 默认值
   */
  async getBackupCron(): Promise<string | null> {
    const dbValue = await this.getSetting('backup_cron');
    if (dbValue !== null) {
      const t = dbValue.trim();
      if (t === '' || t === 'false' || t === 'off' || t === '0') return null;
      return t;
    }
    // fallback env var
    const v = process.env.DB_BACKUP_CRON;
    if (v === undefined || v === null) return DEFAULT_BACKUP_CRON;
    const t = v.trim();
    if (t === '' || t === 'false' || t === 'off' || t === '0') return null;
    return t;
  }

  /**
   * 更新备份目录（校验路径可写）
   */
  async setBackupDir(dir: string): Promise<{ success: boolean; error?: string }> {
    const resolved = path.resolve(dir);
    try {
      fs.mkdirSync(resolved, { recursive: true });
      const testFile = path.join(resolved, '.write-test');
      fs.writeFileSync(testFile, '');
      fs.unlinkSync(testFile);
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      return { success: false, error: `目录不可写: ${msg}` };
    }
    await this.setSetting('backup_dir', resolved);
    return { success: true };
  }

  /**
   * 更新备份 cron 表达式（校验格式合法）
   */
  async setBackupCron(cronExpr: string): Promise<{ success: boolean; error?: string }> {
    const t = cronExpr.trim();
    // 允许关闭：空、false、off、0
    if (t === '' || t === 'false' || t === 'off' || t === '0') {
      await this.setSetting('backup_cron', t);
      return { success: true };
    }
    if (!cron.validate(t)) {
      return { success: false, error: 'Cron 表达式格式无效，请使用标准 5 段格式（分 时 日 月 周）' };
    }
    await this.setSetting('backup_cron', t);
    return { success: true };
  }

  /**
   * 启动时初始化默认设置（仅在 key 不存在时插入）
   */
  async initDefaultSettings(): Promise<void> {
    try {
      const existingDir = await prisma.appSetting.findUnique({ where: { key: 'backup_dir' } });
      if (!existingDir) {
        await prisma.appSetting.create({
          data: {
            key: 'backup_dir',
            value: DEFAULT_BACKUP_DIR,
            updatedAt: BigInt(Date.now()),
          },
        });
        logger.info({ msg: '[AppSetting] 已初始化默认备份目录', backupDir: DEFAULT_BACKUP_DIR });
      }

      const existingCron = await prisma.appSetting.findUnique({ where: { key: 'backup_cron' } });
      if (!existingCron) {
        await prisma.appSetting.create({
          data: {
            key: 'backup_cron',
            value: DEFAULT_BACKUP_CRON,
            updatedAt: BigInt(Date.now()),
          },
        });
        logger.info({ msg: '[AppSetting] 已初始化默认备份定时', backupCron: DEFAULT_BACKUP_CRON });
      }
    } catch (error) {
      logger.error({ msg: '[AppSetting] 初始化默认设置失败', error: error instanceof Error ? error.message : String(error) });
    }
  }
}

export const appSettingService = new AppSettingService();

File: python-backend/app/services/app_setting_service.py
"""App setting service – async SQLAlchemy port of AppSettingService.ts."""

from __future__ import annotations

import os
import re
import time
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.app_setting import AppSetting

_CRON_RE = re.compile(r"^(\S+\s+){4}\S+$")
_DISABLE_VALUES = {"", "false", "off", "0"}


class AppSettingService:
    """Key-value application settings stored in DB."""

    # ------------------------------------------------------------------
    # primitives
    # ------------------------------------------------------------------

    @staticmethod
    async def get_setting(db: AsyncSession, key: str) -> str | None:
        stmt = select(AppSetting).where(AppSetting.key == key)
        result = await db.execute(stmt)
        row = result.scalar_one_or_none()
        return row.value if row is not None else None

    @staticmethod
    async def set_setting(db: AsyncSession, key: str, value: str) -> None:
        now = int(time.time() * 1000)
        stmt = select(AppSetting).where(AppSetting.key == key)
        result = await db.execute(stmt)
        existing = result.scalar_one_or_none()

        if existing is not None:
            existing.value = value
            existing.updated_at = now
        else:
            db.add(AppSetting(key=key, value=value, updated_at=now))
        await db.flush()

    # ------------------------------------------------------------------
    # backup_dir
    # ------------------------------------------------------------------

    @staticmethod
    def _default_backup_dir() -> str:
        from app.core.config import settings

        env_dir = settings.DB_BACKUP_DIR
        return env_dir.strip() if env_dir and env_dir.strip() else os.path.join(os.getcwd(), "db-backup")

    @staticmethod
    async def get_backup_dir(db: AsyncSession) -> str:
        db_value = await AppSettingService.get_setting(db, "backup_dir")
        if db_value and db_value.strip():
            return db_value.strip()
        return AppSettingService._default_backup_dir()

    @staticmethod
    async def set_backup_dir(db: AsyncSession, dir_path: str) -> dict:
        resolved = str(Path(dir_path).resolve())
        try:
            os.makedirs(resolved, exist_ok=True)
            test_file = os.path.join(resolved, ".write-test")
            with open(test_file, "w") as f:
                f.write("")
            os.remove(test_file)
        except Exception as exc:
            return {"success": False, "error": f"目录不可写: {exc}"}
        await AppSettingService.set_setting(db, "backup_dir", resolved)
        return {"success": True}

    # ------------------------------------------------------------------
    # backup_cron
    # ------------------------------------------------------------------

    _DEFAULT_BACKUP_CRON = "0 19 * * 5"

    @staticmethod
    async def get_backup_cron(db: AsyncSession) -> str | None:
        db_value = await AppSettingService.get_setting(db, "backup_cron")
        if db_value is not None:
            t = db_value.strip()
            if t in _DISABLE_VALUES:
                return None
            return t
        # fallback to env
        from app.core.config import settings

        env_val = settings.DB_BACKUP_CRON
        if env_val is None:
            return AppSettingService._DEFAULT_BACKUP_CRON
        t = env_val.strip()
        if t in _DISABLE_VALUES:
            return None
        return t

    @staticmethod
    async def set_backup_cron(db: AsyncSession, cron_expr: str) -> dict:
        t = cron_expr.strip()
        if t in _DISABLE_VALUES:
            await AppSettingService.set_setting(db, "backup_cron", t)
            return {"success": True}
        if not _CRON_RE.match(t):
            return {
                "success": False,
                "error": "Cron 表达式格式无效，请使用标准 5 段格式（分 时 日 月 周）",
            }
        await AppSettingService.set_setting(db, "backup_cron", t)
        return {"success": True}

    # ------------------------------------------------------------------
    # init
    # ------------------------------------------------------------------

    @staticmethod
    async def init_default_settings(db: AsyncSession) -> None:
        try:
            existing_dir = await AppSettingService.get_setting(db, "backup_dir")
            if existing_dir is None:
                db.add(
                    AppSetting(
                        key="backup_dir",
                        value=AppSettingService._default_backup_dir(),
                        updated_at=int(time.time() * 1000),
                    )
                )
                await db.flush()

            existing_cron = await AppSettingService.get_setting(db, "backup_cron")
            if existing_cron is None:
                db.add(
                    AppSetting(
                        key="backup_cron",
                        value=AppSettingService._DEFAULT_BACKUP_CRON,
                        updated_at=int(time.time() * 1000),
                    )
                )
                await db.flush()
        except Exception:
            pass  # swallow – matches TS behaviour of logging and continuing
